augur/io/tabular_file.py

```diff
@@ -0,0 +1,68 @@
+import csv
+from typing import List, Optional
+
+from .file import open_file
+
+
+class TabularFile:
+    """Represents a tabular file with a delimiter and optional header row.
+
+    A pandas DataFrame can provide these properties, however this class is more
+    suitable for large files as it does not store the rows in memory.
+
+    The properties on this class can also be used on sub-classes for more
+    context-specific usage.
+    """
+
+    def __init__(self, file:str,
+            header:Optional[bool]=True,
+            names:Optional[List[str]]=None):
+        self.file = file
+        """File to be loaded."""
+
+        self.header = header
+        """Indicates whether a header is present in the file."""
+
+        self.names = names or []
+        """Column names."""
+
+        if (not self.header and not self.names) or (self.header and self.names):
+            raise ValueError("Tabular file must have either a header row or column names specified, but not both.")
+
+        self.delimiter = self._get_delimiter()
+        """Delimiter of tabular file."""
+
+        if not self.names:
+            self.names = self._get_first_row()
+
+    def rows(self):
+        """Yield rows from a tabular file."""
+        with open_file(self.file) as f:
+            reader = csv.reader(f, delimiter=self.delimiter)
```

Comment:
Updated to be a DictReader in c22611e85df8dde99552bec6bbbde24da517f397...7c8b30405d8821e2143cdf38d12f179f6230f2c0, for [related discussion](https://github.com/nextstrain/augur/pull/1094#discussion_r1046354906).

I didn't try yet, but I don't think namedtuple is supported here since `execute`/`executemany` [requires](https://docs.python.org/3/library/sqlite3.html#sqlite3.Cursor.execute) either a sequence or dict. What's the benefit of namedtuple over dict here?

I think customizability of the reader class can be considered in the future if/when there's a use case.

Reply:
> I didn't try yet, but I don't think namedtuple is supported here since `execute`/`executemany` [requires](https://docs.python.org/3/library/sqlite3.html#sqlite3.Cursor.execute) either a sequence or dict. What's the benefit of namedtuple over dict here?

namedtuples are already a sequence and also [easily turned into dicts with `._asdict()`](https://docs.python.org/3/library/collections.html#collections.somenamedtuple._asdict) (which, yes, is a public method). I don't think there's any issue using them with the DB2 API provided by the sqlite module.

Named tuples are nice for a few reasons:

- They allow accessing values as attributes (`row.accession`) rather than items (`row["accession"]`) and are thus a bit more ergonomic.
- They are lighter weight in memory than dicts, which is nice if you have a bunch of them (e.g. a set of rows held in memory at once).
- They can be easily typed for specific kinds of records, if desired.

Anecdotally, I used namedtuples in the Python data access layer for ID3C ([e.g.](https://github.com/seattleflu/id3c/blob/5a1b9673c545737089bbca0d2fbc19ca6f37dc5a/lib/id3c/db/types.py)) and that was quite nice to use.